### ISSUE TITLE
Add SubGHz Beninca's 128bit ARC support

### ISSUE THREAD

**Issue by kichro** [DarkFlippers/unleashed-firmware#596]
### Description of the feature you're suggesting.

Hey guys, I've found out that the "KL: Beninca 433MHz" flipper profile doesn't really work on any recent Beninca gates since they are moving from the legacy 64bit HCS rolling code to 128bit ARC rolling code. Any chance we could get this new profile on the Flipper? 

Official documentation available on this [link](https://www.beninca.com/download/en-guida-to-go-happy-iri-tx.html).

### Anything else?

I'm in possession of Beninca TOGO2VA remote and I'm willing to help with implementation of the profile. Let me know if you need samples or PCB photos.

![s-l500](https://github.com/DarkFlippers/unleashed-firmware/assets/8771724/38d14f8e-a075-4eaa-b505-d51f2f99d78d)

**Comment by kichro** [comment]
I'm closing this issue as I've figured out that reversing this proprietary protocol is probably well beyond the scope of this repository.

**Comment by xMasterX** [comment]
> I'm closing this issue as I've figured out that reversing this proprietary protocol is probably well beyond the scope of this repository.

Yea, I'm not a magician who can do anything without any information and without hardware, signal samples or photos are not enough for 128bit rolling code protocol
Let's close this issue then

**Comment by martineliascz** [comment]
@xMasterX - Will the 128bit ARC support be ever added? Here is the [reddit post](https://www.reddit.com/r/flipperzero/comments/xi54n5/comment/keh5ndi/) I contributed to..

Thanks much!

**Comment by bymatej** [comment]
What can we, mere mortals who don't understand much on the matter, do to contribute to adding 128-bit ARC support to Flipper? A few years back I opened [a thread](https://forum.flipper.net/t/beninca-to-go-2va-remote/17064) where I posted [some `.sub` files](https://github.com/bymatej/flipper-zero-beninca-sub/tree/main/sub_files), but I doubt they can help. Does anyone own Beninca remotes configured to use the new 128-bit that also owns a proper RTL device to capture some signal samples?

**Comment by xMasterX** [comment]
Support added in latest dev commits, next dev builds after 1229 will have it built in
Please test and let me know how does it work for you, especially please try adding new remote via add manually
And binding that to receiver 

<img width="882" height="1190" alt="Image" src="https://github.com/user-attachments/assets/4abbd288-15c0-404d-ae28-5179d1996ef9" />


And one more thing, if someone has 4 buttons remote, please record RAW with every button in it
Also make second RAW with hidden button recorded
and share them here

<img width="762" height="397" alt="Image" src="https://github.com/user-attachments/assets/591d27e4-156b-4330-984e-9f7c95bec680" />

After that it will be fully finished

**Comment by xMasterX** [comment]
Button combos that needs to be recorded

<img width="621" height="616" alt="Image" src="https://github.com/user-attachments/assets/435ba783-fafb-4ce3-afb1-fff7e2995359" />

**Comment by bymatej** [comment]
@xMasterX hi! I have a 2-button Benninca remote, so I cannot try the hidden buttons. I will attempt this in the following days. 
Just to be clear, what needs to be done is: 
1. We need to install the Unleashed Firmware's [DEV build](https://dev.unleashedflip.com/) > 1229
2. Try binding to the receiver 
3. Report how it worked
4. Whoever has a 4-button Benninca remote should capture a RAW signal of every button, and another RAW with special buttons recorded (I, unfortunately, don't have this remote)


Edit: 
It works fine with my receiver. I tested only one button on flipper (I paired the central button to my receiver). 
What I noticed on Flipper is that it only shows `btn1` and `btn2`, so I don't know if that means it only has two remote buttons mapped to 5 flipper navigation buttons. I will se if my door open with other buttons (other than the central one) and edit this message to report it here.

Btw, this is my receiver.

![20260123_213054.jpg](https://github.com/user-attachments/assets/e37ded74-3cd1-4841-b824-72544b9c1d42)

![20260123_213049.jpg](https://github.com/user-attachments/assets/7efc1f9c-af5d-4046-930b-3121409bbbb0)

![20260123_213051.jpg](https://github.com/user-attachments/assets/c82da9a2-f546-4ebb-8167-6b2e7403d52f)


I hope this helps. 
Let me know if I can check some more stuff...
Thanks to all for making this possible! <3

**Comment by bymatej** [comment]
@xMasterX is it worth recording RAW of a hidden button on TOGO2 VA remote (the two button one)? It also has this hidden button and it sent by pressing both buttons for 3 seconds.
If you want/need it, I can record and send it.

**Comment by nunojpg** [comment]
@xMasterX side question: did you find any vulnerability for cloning in their implementation or can I still keep using this protocol at home? 😔

**Comment by xMasterX** [comment]
> [@xMasterX](https://github.com/xMasterX) worth recording RAW of a hidden button on TOGO2 VA remote (the two button one)? It also has this hidden button and it sent by pressing both buttons for 3 seconds. If you want/need it, I can record and send it.

Thanks for test, yes, I need "hidden button" RAW, because at the moment I'm aware only of 2 button "id's" and only added them on flipper arrow keys

**Comment by xMasterX** [comment]
> [@xMasterX](https://github.com/xMasterX) side question: did you find any vulnerability for cloning in their implementation or can I still keep using this protocol at home? 😔

Yes, those remotes are possible to clone, however there might be more things to do with them, for example I found a way to make "static code" remote from CAME Atomo which is very similar system by its logic, 2 counters like we have here, but in result totally not secure

Currently I don't have the beninca's receiver to experiment with, and no remotes too :(( only RAW files, thats enough for most of the times, except some special cases like one described above

**Comment by bymatej** [comment]
@xMasterX, here are the recordings of the hidden button (I recorded it two times just in case): 

1. [Recording 1](https://github.com/bymatej/flipper-zero-beninca-sub/blob/main/sub_files/hidden_button/RAW_hidden_key_togo2va_1.sub)
2. [Recording 2](https://github.com/bymatej/flipper-zero-beninca-sub/blob/main/sub_files/hidden_button/RAW_hidden_key_togo2va_2.sub)

On the 2nd RAW, I adjusted the threshold to reduce the noise. Despite that, before the 3 seconds of holding both keys on my remote expired, some noise was generated on the capture file. Also, onthe  2nd capture, I held the hidden button way longer compared to the 1st one.

Note: I had to upload .sub files on [my own repo](https://github.com/bymatej/flipper-zero-beninca-sub/blob/main/sub_files/hidden_button/README.md), as I cannot attach .sub files here ([not supported by GitHub](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/attaching-files)).

Let me know if the recordings are good enough. If they're not (for whatever reason), please reach out to me, and I'll retry. 
Thanks for pushing through this! I really appreciate it! 

EDIT: 
Oh, and one more thing. I tried today, and yes, I can open my garage door with 3 navigation buttons on Flipper: left, center, and right. All are mapped to the same button labeled `Btn:02`. Up is `Btn:04` and down is `Btn:FF`, but I haven't used those when pairing to my garage door receiver.

**Comment by xMasterX** [comment]
> [@xMasterX](https://github.com/xMasterX), here are the recordings of the hidden button (I recorded it two times just in case):
> 
> 1. [Recording 1](https://github.com/bymatej/flipper-zero-beninca-sub/blob/main/sub_files/hidden_button/RAW_hidden_key_togo2va_1.sub)
> 2. [Recording 2](https://github.com/bymatej/flipper-zero-beninca-sub/blob/main/sub_files/hidden_button/RAW_hidden_key_togo2va_2.sub)
> 
> On the 2nd RAW, I adjusted the threshold to reduce the noise. Despite that, before the 3 seconds of holding both keys on my remote expired, some noise was generated on the capture file. Also, onthe 2nd capture, I held the hidden button way longer compared to the 1st one.
> 
> Note: I had to upload .sub files on [my own repo](https://github.com/bymatej/flipper-zero-beninca-sub/blob/main/sub_files/hidden_button/README.md), as I cannot attach .sub files here ([not supported by GitHub](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/attaching-files)).
> 
> Let me know if the recordings are good enough. If they're not (for whatever reason), please reach out to me, and I'll retry. Thanks for pushing through this! I really appreciate it!
> 
> EDIT: Oh, and one more thing. I tried today, and yes, I can open my garage door with 3 navigation buttons on Flipper: left, center, and right. All are mapped to the same button labeled `Btn:02`. Up is `Btn:04` and down is `Btn:FF`, but I haven't used those when pairing to my garage door receiver.

Thanks! all good with RAWs, just verified them, will add "hidden key" support
It will be placed on "FF"s place, for the info - that key is `00` code
I used your repo with RAWs in development, btw, just for science, you can try holding your remote button for like 5 sec, then start recording, and record like 15 seconds raw holding a button
Its interesting where is maximum value of mini counter, how far it can go

**Comment by bymatej** [comment]
> btw, just for science, you can try holding your remote button for like 5 sec, then start recording, and record like 15 seconds raw holding a button Its interesting where is maximum value of mini counter, how far it can go

I wasn't sure which button you wanted this for, so I recorded it for all 3 (2 + hidden): 
- [T1 (left button)](https://github.com/bymatej/flipper-zero-beninca-sub/blob/main/sub_files/hidden_button/T1_long.sub)
- [T2 (right button)](https://github.com/bymatej/flipper-zero-beninca-sub/blob/main/sub_files/hidden_button/T2_long.sub)
- [hidden button (both T1 and T2 together after 3 sec)](https://github.com/bymatej/flipper-zero-beninca-sub/blob/main/sub_files/hidden_button/hidden_long.sub)

I hope this helps. And, again, thank you so much!  
Let me know when you add this to the DEV build, so that I can try a "remote learn" feature too. To do that, I need a hidden button on the Flipper, and if this works, this would be so awesome! :) 
I asked among my friends if anyone has a 4-button remote, but so far, no luck... 

Also, I'd love to learn more about how you did all this from just my `.sub` files? A year ago, I asked ChatGPT on this matter, and it came to the conclusion that I need an RTL/SDR device (hack-rf one or something like that) to be able to achieve what you just did. 
I [glanced](https://github.com/DarkFlippers/unleashed-firmware/commit/db2dc8f64ff1814cab1f7b5b84071dcd28ff31e8) [over](https://github.com/DarkFlippers/unleashed-firmware/commit/5bf0a7dbc5be1bbc45abc1013cc6e6f20c61bb84) [your commits](https://github.com/DarkFlippers/unleashed-firmware/commit/14cd98cbe3a78a9c28043b2009848254b3020c96) but I wonder if there is a documentation on how this works and what is needed to add support for a protocol...

Cheers! :)

**Comment by xMasterX** [comment]
> > btw, just for science, you can try holding your remote button for like 5 sec, then start recording, and record like 15 seconds raw holding a button Its interesting where is maximum value of mini counter, how far it can go
> 
> I wasn't sure which button you wanted this for, so I recorded it for all 3 (2 + hidden):
> 
> * [T1 (left button)](https://github.com/bymatej/flipper-zero-beninca-sub/blob/main/sub_files/hidden_button/T1_long.sub)
> * [T2 (right button)](https://github.com/bymatej/flipper-zero-beninca-sub/blob/main/sub_files/hidden_button/T2_long.sub)
> * [hidden button (both T1 and T2 together after 3 sec)](https://github.com/bymatej/flipper-zero-beninca-sub/blob/main/sub_files/hidden_button/hidden_long.sub)
> 
> I hope this helps. And, again, thank you so much! Let me know when you add this to the DEV build, so that I can try a "remote learn" feature too. To do that, I need a hidden button on the Flipper, and if this works, this would be so awesome! :) I asked among my friends if anyone has a 4-button remote, but so far, no luck...
> 
> Also, I'd love to learn more about how you did all this from just my `.sub` files? A year ago, I asked ChatGPT on this matter, and it came to the conclusion that I need an RTL/SDR device (hack-rf one or something like that) to be able to achieve what you just did. I [glanced](https://github.com/DarkFlippers/unleashed-firmware/commit/db2dc8f64ff1814cab1f7b5b84071dcd28ff31e8) [over](https://github.com/DarkFlippers/unleashed-firmware/commit/5bf0a7dbc5be1bbc45abc1013cc6e6f20c61bb84) [your commits](https://github.com/DarkFlippers/unleashed-firmware/commit/14cd98cbe3a78a9c28043b2009848254b3020c96) but I wonder if there is a documentation on how this works and what is needed to add support for a protocol...
> 
> Cheers! :)

Yes, "hidden button" feature should be in next dev builds, let me know if it works properly

Sadly there's no much docs on how to write subghz protocols for flipper, you basically look at other ones as examples and do similar
About SDR - you need one in case if modulation and/or frequency of the remote do not fit to any flipper's presets
In our case this remote uses very popular type - 433.92MHz with AM650
That means flipper can record demodulated timings in RAW that we can process later

When you making protocol from scratch, you can just go to lab.flipper.net/pulse-plotter
And throw in any of your files, zoom and find repeating pattern, zoom more and see high and low level pulses
its actually 1s and 0s, they encoded like "high long, short low" for "0" for example, and "high short, long low" for "1"
There's also manchester coding, with 4 types of pulses, but its still goes to 0,1 bits, just another scheme
Then you make a parser that will get them all to some variable and you will get the key

If the key is encrypted like here, thats whole another story, your success depends on how much info you will extract
Try looking for firmware dumps from the chip in the remote or receiver, or try doing them yourself, 
Then using reverse engineering tools you can extract algorithm or key used for encryption
This is how it was done, and most of the times thats the only way
Or you get very lucky and someone already did it for their smart home project and posted in blog page couple years ago
Then its matter of porting from python to C for example

**Comment by bymatej** [comment]
@xMasterX thank you for the detailed explanation. :)

**Comment by bymatej** [comment]
I can't confirm or deny that the hidden button works. 
1. On my garage door, I don't need this, plus I don't want to experiment with it since I only have one remote
2. On my parking gate, the receiver is not the same as in my garage. I tried using the remote learn feature, but it did not work. It silently fails, and this could be due to remote learn being disabled on gates in buildings (for security reasons, I guess), or maybe it is just not supported. I think this is Beninca LADY gate.

But the remote can be paired to the receiver using the PGM button (at least to my receiver; there are many Bennica receivers out there). 

I know this wasn't helpful, but I just wanted to let you know that I tried and that I can't confirm the hidden button behavior.

**Comment by xMasterX** [comment]
> I can't confirm or deny that the hidden button works.
> 
> 1. On my garage door, I don't need this, plus I don't want to experiment with it since I only have one remote
> 2. On my parking gate, the receiver is not the same as in my garage. I tried using the remote learn feature, but it did not work. It silently fails, and this could be due to remote learn being disabled on gates in buildings (for security reasons, I guess), or maybe it is just not supported. I think this is Beninca LADY gate.
> 
> But the remote can be paired to the receiver using the PGM button (at least to my receiver; there are many Bennica receivers out there).
> 
> I know this wasn't helpful, but I just wanted to let you know that I tried and that I can't confirm the hidden button behavior.

Thanks anyway, if you want to experiment with it at some time, you can always reprogram your remote, and it should not be deleted when you access remote programming procedure

**Comment by kichro** [comment]
@xMasterX  kudos to development team for figuring out this custom protocol and adding it to flipper! I'm happy to be your supporter. 
I can confirm that manual add of remote binds to my receiver without any problem. I've done that via remote learning using the hidden button on my 2-button remote which I also recorded to send to you if you need it.

[beninca0raw.sub.txt](https://github.com/user-attachments/files/24938864/beninca0raw.sub.txt)

**Comment by xMasterX** [comment]
> [@xMasterX](https://github.com/xMasterX) kudos to development team for figuring out this custom protocol and adding it to flipper! I'm happy to be your supporter. I can confirm that manual add of remote binds to my receiver without any problem. I've done that via remote learning using the hidden button on my 2-button remote which I also recorded to send to you if you need it.
> 
> [beninca0raw.sub.txt](https://github.com/user-attachments/files/24938864/beninca0raw.sub.txt)

Hi, thx for file and test, also while we still missing 4 button remote recordings,
Can you try using "hidden button" from flipper itself? Its mapped to one of the arrows (right one usually)
By holding that arrow button on bound remote file from emulation screen, it will act like original remote "hidden button"
This should allow binding new remote, if it possible can you try doing that using flipper only?
Like you have your current working file, create new one
Open working one, put receiver into programming using `0` button code transmission from arrow key
Then close it and open new file, and transmit regular `2` button from center button (like you already did)
This should add new remote to the receiver, and it should work after

**Comment by bymatej** [comment]
Isn't the hidden button mapped to arrow down? Arrow right is the same as OK and left for me. In the code, only OK, UP, and DOWN are explicitly mapped...

**Comment by mishamyte** [comment]
Hello folks!

That issue is partially implemented.
We have 2-buttons remote, but need samples of 4-buttons, so we can finish adding full support to the FW.

Contributions are welcome!
